sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/instrumentation/ServiceBusTracer.java

```diff
@@ -0,0 +1,406 @@
+
+        Context eventSpanContext = tracer.start("ServiceBus.message", newMessageContext, ProcessKind.MESSAGE);
+        Optional<Object> traceparentOpt = eventSpanContext.getData(DIAGNOSTIC_ID_KEY);
+
+        if (traceparentOpt.isPresent()) {
+            serviceBusMessage.getApplicationProperties().put(DIAGNOSTIC_ID_KEY, traceparentOpt.get().toString());
+            serviceBusMessage.getApplicationProperties().put(TRACEPARENT_KEY, traceparentOpt.get().toString());
+
+            endSpan(null, eventSpanContext, null);
+
+            Optional<Object> spanContext = eventSpanContext.getData(SPAN_CONTEXT_KEY);
+            if (spanContext.isPresent()) {
+                serviceBusMessage.addContext(SPAN_CONTEXT_KEY, spanContext.get());
+            }
+        }
+    }
+
+    /**
+     * Instruments peek or receiveDeferred that return a single message. Creates a single span, does not report any metrics
+     */
+    public Mono<ServiceBusReceivedMessage> traceManagementReceive(String spanName, Mono<ServiceBusReceivedMessage> publisher,
+        Function<ServiceBusReceivedMessage, Context> getMessageContext) {
+        if (tracer != null) {
+            AtomicLong startTime = new AtomicLong();
+            AtomicReference<ServiceBusReceivedMessage> message = new AtomicReference<>();
+            return publisher.doOnEach(signal -> {
+                if (signal.hasValue()) {
+                    message.set(signal.get());
+                }
+
+                if (signal.isOnComplete() || signal.isOnError()) {
+                    ServiceBusReceivedMessage msg = message.get();
+                    Context messageContext = msg == null ? null : getMessageContext.apply(msg);
+
+                    Context span = startSpanWithLink(spanName, msg, messageContext, new Context(START_TIME_KEY, startTime.get()));
+                    endSpan(null, span, null);
+                }
+            })
+            .doOnSubscribe(s -> {
+                startTime.set(Instant.now().toEpochMilli());
+            });
+        }
+        return publisher;
+    }
+
+
+    /**
+     * Traces receive, peek, receiveDeferred that return a flux of messages, but has a limited lifetime - such as sync receive case or peek many
+     * or receive deferred.
+     *
+     * Don't use it for async receive when Flux has unknown lifetime!!!
+     *
+     * Creates a single span with links to each message being received.
+     */
+    public Flux<ServiceBusReceivedMessage> traceSyncReceive(String name, Flux<ServiceBusReceivedMessage> messages) {
+        if (tracer != null) {
+            AtomicLong startTime = new AtomicLong();
+            List<ServiceBusReceivedMessage> messageList = new ArrayList<>();
+            return messages
+                .doOnEach(signal -> {
+                    if (signal.hasValue()) {
+                        ServiceBusReceivedMessage message = signal.get();
+                        messageList.add(message);
```

Comment:
Please excuse my lack of understanding in tracing - Is there a way to achieve this without storing all messages? Like creating a spanBuilder on subscribe, addLink as we receive a message, and finally, when flux (bounded by max-message) terminates, we start and end span?

Reply:
You're absolutely right!

Reply:
thank you so much!